### PR TITLE
Add Floyd's Cycle Detection Algorithm

### PR DESCRIPTION
**Pull request by aqib-m31**
### Describe your change:

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Add or change doctests? -- Note: Please avoid changing both code and tests in a single pull request.
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [x] If this pull request resolves one or more open issues then the description above includes the issue number(s) with a [closing keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue): "Fixes #ISSUE-NUMBER".

### CODE REVIEW

File: data_structures/linked_list/floyds_cycle_detection.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
+"""
+Floyd's cycle detection algorithm is a popular algorithm used to detect cycles
+in a linked list. It uses two pointers, a slow pointer and a fast pointer,
+to traverse the linked list. The slow pointer moves one node at a time while the fast
+pointer moves two nodes at a time. If there is a cycle in the linked list,
+the fast pointer will eventually catch up to the slow pointer and they will
+meet at the same node. If there is no cycle, the fast pointer will reach the end of
+the linked list and the algorithm will terminate.
+
+For more information: https://en.wikipedia.org/wiki/Cycle_detection#Floyd's_Tortoise_and_Hare
```
Comment (tianyizheng02) [inline data_structures/linked_list/floyds_cycle_detection.py:10]:
```suggestion
For more information: https://en.wikipedia.org/wiki/Cycle_detection#Floyd's_tortoise_and_hare
```
The URL is case-sensitive

File: data_structures/linked_list/floyds_cycle_detection.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+
+@dataclass
+class Node:
+    """
+    A class representing a node in a singly linked list.
+    """
+
+    data: Any
+    next_node: "Node" | None = None
```
Comment (tianyizheng02) [inline data_structures/linked_list/floyds_cycle_detection.py:25]:
```suggestion
    next_node: Self | None = None
```
This line is causing the build to fail. As of Python 3.11, you can use the `Self` type to annotate a variable with the name of the enclosing class. I believe this should fix the build error.

**Review by tianyizheng02 — APPROVED** [review]

File: data_structures/linked_list/floyds_cycle_detection.py (hunk 2 above)
Comment (aqib-m31) [reply data_structures/linked_list/floyds_cycle_detection.py:25]: Oh! Let me try again!
Actually I tried to run the tests on my system. But I'm getting some errors - I couldn't fix any of those for 4 hrs. I'm getting one error if I use power shell and other error if I use WSL.

Comment (aqib-m31) [reply data_structures/linked_list/floyds_cycle_detection.py:25]: @tianyizheng02 I delete the repo, I'll try again!
But I am still not able to understand why am I getting errors on my system, I also posted the same on discord.
